Er/ct 2969 breaking contract warning

emmyoop opened the pull request:
resolves #8282 
resolves #8384
[docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose) dbt-labs/docs.getdbt.com/#  


### Problem

1. Error message is not clear
2. dbt throws an error when it's better to throw a warning when a contracted model is unversioned

### Solution

1. Reformat messages
2. Conditionally throw a warning for unversioned models with the ability to mark that warning to be an error.

### Checklist

- [ ] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me  
- [ ] I have run this code in development and it appears to resolve the stated issue  
- [ ] This PR includes tests, or tests are not required/relevant for this PR
- [ ] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX

emmyoop commented:
### Warning output
```
(env-version-models) emily@Emily-Rockman basic-dbt % dbt run --models state:modified --state ./state
13:07:58  Running with dbt=1.7.0-a1
13:07:58  Registered adapter: postgres=1.7.0-a1
13:07:59  Found 17 models, 1 seed, 11 tests, 1 source, 3 exposures, 1 metric, 468 macros, 2 groups, 1 semantic model
13:07:59  Breaking Change to Unversioned Contract for contracted_model (models/contracted_models/contracted_model.sql)
While comparing to previous project state, dbt detected a breaking change to an unversioned model.
  - Contract enforcement was removed: Previously, this model's configuration included contract: {enforced: true}. It is no longer configured to enforce its contract, and this is a breaking change.
  - Columns were removed:
    - color
    - date_day
  - Enforced column level constraints were removed:
    - id (ConstraintType.not_null)
    - id (ConstraintType.primary_key)
  - Enforced model level constraints were removed:
    - ConstraintType.check -> ['id']
  - Materialization changed with enforced constraints:
    - table -> view

```

### Error Output
```
(env-version-models) emily@Emily-Rockman basic-dbt % dbt run --models state:modified --state ./state
13:07:16  Running with dbt=1.7.0-a1
13:07:16  Registered adapter: postgres=1.7.0-a1
13:07:17  Found 17 models, 1 seed, 11 tests, 1 source, 3 exposures, 1 metric, 468 macros, 2 groups, 1 semantic model
13:07:17  Encountered an error:
Breaking Change to Contract Error in model contracted_model_versioned (models/contracted_versions_models copy/contracted_model_versioned.sql)
  While comparing to previous project state, dbt detected a breaking change to an enforced contract.
    - Columns were removed:
      - color
      - date_day
    - Enforced column level constraints were removed:
      - id (ConstraintType.not_null)
      - id (ConstraintType.primary_key)
    - Materialization changed with enforced constraints:
      - table -> view
  Consider making an additive (non-breaking) change instead, if possible.
  Otherwise, create a new model version: https://docs.getdbt.com/docs/collaborate/govern/model-versions
```

peterallenwebb reviewed (approved):
Looks good.

emmyoop commented:
@graciegoheen or @jtcohen6 can I get feedback on the reworked error/warning messages?  Examples posted above.

graciegoheen commented:
This is looking so much better! A few cosmetic / nit-picks on wording (none are blocking):
-  Is there a specific reason for capitalizing "Breaking Change", "Unversioned Contract", "Contract Error"?
- `While comparing to previous project state, dbt detected a breaking change to an unversioned model.` Thoughts on changing the wording on this to something like `While comparing to previous project state, dbt detected a breaking change to an unversioned, previously contracted model.` - this error wouldn't pop up for a plain 'ole unversioned model, so I think it would be nice to callout that it _used to be contracted_ in the initial error message.
-   `Contract enforcement was removed: Previously, this model's configuration included contract: {enforced: true}. It is no longer configured to enforce its contract, and this is a breaking change.` I think it would be more human readable to say something like `Previously, this model had an enforced contract...` rather than including the actual config block `contract: {enforced: true}`.
- +1 to @jtcohen6's suggestions on the constraints suggestions above
- `Breaking Change to Contract Error in model contracted_model_versioned` The `Error` seems redundant here since this is an error message - is there a reason for it? I would expect to see something like `Breaking change to contracted model - contracted_model_versioned` instead.
- `Breaking Change to Unversioned Contract for contracted_model` Similarly to above, I might suggest making this clearer - something like, `Breaking change to contracted, unversioned model - contracted_model`.
- When are we outputting a warning or unversioned models vs. an error? Would we ever recommend that they version their model if it's contracted but not versioned? I suppose in this case, the person is actually removing the contract entirely...

@jtcohen6 @emmyoop feel free to disagree with any of the above! Curious to hear your thoughts

emmyoop commented:
New messages:

### Warning output for unversioned models
```
(env-version-models) emily@Emily-Rockman basic-dbt % dbt run --models state:modified --state ./state
16:17:24  Running with dbt=1.7.0-a1
16:17:24  Registered adapter: postgres=1.7.0-a1
16:17:24  Found 18 models, 1 seed, 11 tests, 1 source, 3 exposures, 1 metric, 468 macros, 2 groups, 1 semantic model
16:17:24  Breaking change to contracted, unversioned model contracted_model (models/contracted_models/contracted_model.sql)
While comparing to previous project state, dbt detected a breaking change to an unversioned model.
  - Contract enforcement was removed: Previously, this model had an enforced contract. It is no longer configured to enforce its contract, and this is a breaking change.
  - Columns were removed:
    - color
  - Columns with data_type changes:
    - date_day (date -> text)
  - Enforced column level constraints were removed:
    - 'not_null' constraint on column id
    - 'primary_key' constraint on column id
  - Enforced model level constraints were removed:
    - 'check' constraint on columns ['id']
  - Materialization changed with enforced constraints:
    - table -> view
```

### Error output for unversioned models
```
(env-version-models) emily@Emily-Rockman basic-dbt % dbt --warn-error run --models state:modified --state ./state
16:17:58  Running with dbt=1.7.0-a1
16:17:58  Registered adapter: postgres=1.7.0-a1
16:17:58  Found 18 models, 1 seed, 11 tests, 1 source, 3 exposures, 1 metric, 468 macros, 2 groups, 1 semantic model
16:17:58  Encountered an error:
Compilation Error
  Breaking change to contracted, unversioned model contracted_model (models/contracted_models/contracted_model.sql)
  While comparing to previous project state, dbt detected a breaking change to an unversioned model.
    - Contract enforcement was removed: Previously, this model had an enforced contract. It is no longer configured to enforce its contract, and this is a breaking change.
    - Columns were removed:
      - color
    - Columns with data_type changes:
      - date_day (date -> text)
    - Enforced column level constraints were removed:
      - 'not_null' constraint on column id
      - 'primary_key' constraint on column id
    - Enforced model level constraints were removed:
      - 'check' constraint on columns ['id']
    - Materialization changed with enforced constraints:
      - table -> view
```


### Error Output for versioned models
```
(env-version-models) emily@Emily-Rockman basic-dbt % dbt run --models state:modified --state ./state
16:20:07  Running with dbt=1.7.0-a1
16:20:07  Registered adapter: postgres=1.7.0-a1
16:20:08  Found 18 models, 1 seed, 11 tests, 1 source, 3 exposures, 1 metric, 468 macros, 2 groups, 1 semantic model
16:20:08  Breaking change to contracted, unversioned model contracted_model (models/contracted_models/contracted_model.sql)
While comparing to previous project state, dbt detected a breaking change to an unversioned model.
  - Contract enforcement was removed: Previously, this model had an enforced contract. It is no longer configured to enforce its contract, and this is a breaking change.
  - Columns were removed:
    - color
  - Columns with data_type changes:
    - date_day (date -> text)
  - Enforced column level constraints were removed:
    - 'not_null' constraint on column id
    - 'primary_key' constraint on column id
  - Enforced model level constraints were removed:
    - 'check' constraint on columns ['id']
  - Materialization changed with enforced constraints:
    - table -> view

```

peterallenwebb commented on core/dbt/contracts/graph/nodes.py:
```diff
@@ -682,11 +683,11 @@ def same_contract(self, old, adapter_type=None) -> bool:
-        enforced_model_constraint_removed: List[
-            Tuple[str, List[str]]
-        ] = []  # constraint_type, columns
+        column_type_changes: List[Dict[str, str]] = []
```
The dict is better than a tuple, but now I'm thinking it would make sense to make this a dataclass called ColumnTypeChange, and analogously for the below, ColumnConstraintRemoval and ModelConstraintRemoval.

peterallenwebb replied:
I don't think it's a huge deal, though so if you want to go ahead anyway, I'm cool with it. Otherwise this changeset looks good.
